CurrencyConverterControl/CurrencyConverterControl.cs

```diff
@@ -0,0 +1,334 @@
+        public ICurrencyDataProvider CurrencyConverter
+        {
+            get { return CurrencyDataProvider; }
+            set { CurrencyDataProvider = value ?? throw new ArgumentNullException(nameof(value)); }
+        }
+
+        public static readonly DependencyProperty CustomStyleProperty =
+            DependencyProperty.Register(nameof(CustomStyle), typeof(Style), typeof(CurrencyConverterControl));
+        /// <summary>
+        /// Define custom styles on the control
+        /// </summary>
+        public Style CustomStyle
+        {
+            get { return (Style)GetValue(CustomStyleProperty); }
+            set { SetValue(CustomStyleProperty, value); }
+        }
+        #endregion
+        public override void OnApplyTemplate()
+        {
+
+            LoadCurrencies();
+            CurrencyDataProvider.GetConvertionRates();
+            _currencyDestination = Template.FindName("CmbDestination", this) as ComboBox;
+            _currencySource = Template.FindName("CmbSource", this) as ComboBox;
+            //To Be done
+            if (_currencyDestination is null || _currencySource is null)
+            { throw new InvalidOperationException("The Combobox in the Template is not found! Please double check!"); }
+            else 
+            {
+                _currencyDestination.Loaded += ComboBox_Loaded;         
+                _currencySource.Loaded += ComboBox_Loaded;
+                _currencyDestination.SelectionChanged += CmbDestination_SelectionChanged;
+                _currencySource.SelectionChanged += CmbSource_SelectionChanged;   
+            }
+                     
+            InputValue = 233.3; //providing a default value            
+            base.OnApplyTemplate();
+        }
+
+        public virtual Task LoadAsync()
+        {
+            // Call the protected LoadCurrencies method
+            LoadCurrencies();
+            return Task.CompletedTask;
+        }
+
+        //Method to load the currencies list
+        private void LoadCurrencies()
+        {
+            try
+            {
+                if (CurrencyList.Any())
+                {
+                    return;
+                }
+                var currencies = CurrencyDataProvider.GetCurrenciesData();
+
+                if (currencies.Any())
+                {
+                    foreach (var currency in currencies)
+                    {
+
+                        CurrencyList.Add(currency);
+                    }
+                }
+
+            }
+            catch (Exception ex) { throw ex; }
+
+        }
+
+        #region Events
+        //Method to set the default value after the ComboBox is loaded
+        private void ComboBox_Loaded(object sender, RoutedEventArgs e)
+        {
+            
+            var comboBox = sender as ComboBox;
+            if (comboBox != null && comboBox.Items.Count > 0 && comboBox.SelectedIndex == -1)
+            {
+                comboBox.SelectedIndex = 0; 
```

Comment:
what if application set source and target currencies before application start? Is that possible? If yes, then default value might be not 0